cli/module_loader.rs

```diff
@@ -900,3 +1028,79 @@ impl ModuleGraphUpdatePermit for WorkerModuleGraphUpdatePermit {
     drop(self.permit); // explicit drop for clarity
   }
 }
+
+#[derive(Debug)]
+struct CliNodeRequireLoader<TGraphContainer: ModuleGraphContainer> {
+  emitter: Arc<Emitter>,
+  fs: Arc<dyn FileSystem>,
+  graph_container: TGraphContainer,
+  in_npm_pkg_checker: Arc<dyn InNpmPackageChecker>,
+  npm_resolver: Arc<dyn CliNpmResolver>,
+}
+
+impl<TGraphContainer: ModuleGraphContainer>
+  CliNodeRequireLoader<TGraphContainer>
+{
+  pub fn new(
+    emitter: Arc<Emitter>,
+    fs: Arc<dyn FileSystem>,
+    graph_container: TGraphContainer,
+    in_npm_pkg_checker: Arc<dyn InNpmPackageChecker>,
+    npm_resolver: Arc<dyn CliNpmResolver>,
+  ) -> Self {
+    Self {
+      emitter,
+      fs,
+      graph_container,
+      in_npm_pkg_checker,
+      npm_resolver,
+    }
+  }
+}
+
+impl<TGraphContainer: ModuleGraphContainer> NodeRequireLoader
+  for CliNodeRequireLoader<TGraphContainer>
+{
+  fn ensure_read_permission<'a>(
+    &self,
+    permissions: &mut dyn deno_runtime::deno_node::NodePermissions,
+    path: &'a Path,
+  ) -> Result<std::borrow::Cow<'a, Path>, AnyError> {
+    if let Ok(url) = deno_path_util::url_from_file_path(path) {
+      // allow reading if it's in the module graph
+      if self.graph_container.graph().get(&url).is_some() {
+        return Ok(std::borrow::Cow::Borrowed(path));
+      }
+    }
+    self.npm_resolver.ensure_read_permission(permissions, path)
+  }
+
+  fn load_text_file_lossy(&self, path: &Path) -> Result<String, AnyError> {
+    // todo(dsherret): use the preloaded module from the graph if available?
```

Comment:
Is this actionable? Maybe open an issue about it?

Reply:
It's maybe a performance optimization, but it also might introduce undesirable behaviour. I'm not sure it's worth tracking in an issue.